lib/stripe/util.rb

```diff
@@ -258,17 +323,71 @@ def self.check_array_of_maps_start_keys!(arr)
         end
       end
     end
+    private_class_method :check_array_of_maps_start_keys!
 
-    # Constant time string comparison to prevent timing attacks
-    # Code borrowed from ActiveSupport
-    def self.secure_compare(a, b)
-      return false unless a.bytesize == b.bytesize
+    # Uses an ANSI escape code to colorize text if it's going to be sent to a
+    # TTY.
+    def self.colorize(val, color, isatty)
+      return val unless isatty
 
-      l = a.unpack "C#{a.bytesize}"
+      mode = 0 # default
+      foreground = 30 + COLOR_CODES.fetch(color)
+      background = 40 + COLOR_CODES.fetch(:default)
 
-      res = 0
-      b.each_byte { |byte| res |= byte ^ l.shift }
-      res == 0
+      "\033[#{mode};#{foreground};#{background}m#{val}\033[0m"
+    end
+    private_class_method :colorize
+
+    # TODO: Make these named required arguments when we drop support for Ruby
+    # 2.0.
+    def self.log_internal(message, data = {}, color: nil, level: nil, out: nil)
+      data_str = data.select { |k,v| !v.nil? }.
+        map { |(k,v)|
+          "%s=%s" % [
+            colorize(k, color, out.isatty),
+            wrap_logfmt_value(v)
+          ]
+        }.join(" ")
+
+      if out.isatty
+        out.puts "%s %s %s" %
+          [colorize(level[0, 4].upcase, color, out.isatty), message, data_str]
+      else
+        out.puts "message=%s level=%s %s" %
+          [wrap_logfmt_value(message), level, data_str]
+      end
+    end
+    private_class_method :log_internal
+
+    def self.titlecase_parts(s)
+      s.split("-").
+        select { |p| p != "" }.
+        map { |p| p[0].upcase + p[1..-1].downcase }.
+        join("-")
+    end
+    private_class_method :titlecase_parts
+
+    # Wraps a value in double quotes if it looks sufficiently complex so that
+    # it can be read by logfmt parsers.
+    def self.wrap_logfmt_value(val)
+      # If value is any kind of number, just allow it to be formatted directly
+      # to a string (this will handle integers or floats).
+      return val if val.is_a?(Numeric)
+
+      # Hopefully val is a string, but protect in case it's not.
+      val = val.to_s
+
+      if %r{\A[\w\-/]+\z} =~ val
```

Comment:
Yeah, it would be nice if there was a better way, but in practice it probably doesn't matter all that much.

To make this a little more performance, I'm going to flip this:

```
\A[\w\-/]+\z
```

In its inverted form:

```
[^\w\-/]
```

Which should be a little faster because we no longer need to match the entire string. If a single non-conforming character is matched, it can return immediately.

Patched in f546ddf.